calendar_resource/models/calendar_event.py

```diff
@@ -0,0 +1,268 @@
+                    fields.Datetime.to_string(interval[1]),
+                )
+            args = {
+                'start': interval[0],
+                'stop': interval[1],
+                'zallday': False,
+                'zduration': 24,
+            }
+            datetimes.append(self._get_display_time(**args))
+        return (2 * linesep).join(datetimes)
+
+    @api.multi
+    def _event_in_past(self):
+        self.ensure_one()
+        stop_datetime = fields.Datetime.from_string(self.stop)
+        now_datetime = datetime.now()
+        return stop_datetime < now_datetime
+
+    @api.multi
+    @api.constrains('resource_ids', 'start', 'stop')
+    def _check_resource_ids_double_book(self):
+
+        for record in self:
+
+            if record._event_in_past():
+                continue
+
+            resources = record.resource_ids.filtered(
+                lambda s: s.allow_double_book is False
+            )
+
+            if not any(resources):
+                continue
+
+            overlaps = self.env['calendar.event'].search([
+                ('id', '!=', record.id),
+                ('start', '<', record.stop),
+                ('stop', '>', record.start),
+            ])
+            for resource in overlaps.mapped(lambda s: s.resource_ids):
+                raise ValidationError(
+                    _(
+                        'The resource, %s, cannot be double-booked '
+                        'with any overlapping meetings or events.',
+                    )
+                    % resource.name,
+                )
+
+    @api.multi
+    @api.constrains('resource_ids', 'categ_ids')
+    def _check_resource_ids_categ_ids(self):
+
+        for record in self:
+
+            if record._event_in_past():
+                continue
+
+            if not record.categ_ids:
+                continue
+
+            for resource in record.resource_ids:
+                categs = record.categ_ids.filtered(
+                    lambda s: s not in resource.allowed_event_types
+                )
+                if categs:
+                    raise ValidationError(
+                        _(
+                            "The resource, '%s', is not allowed in the "
+                            "following event types: \n%s",
+                        )
+                        % (
+                            resource.name,
+                            ', '.join([
+                                categ.name for categ in categs
+                            ]),
+                        )
+                    )
+
+    @api.multi
+    @api.constrains('resource_ids', 'start', 'stop')
+    def _check_resource_ids_leaves(self):
+
+        for record in self:
+
+            if record._event_in_past():
+                continue
+
+            for resource in record.resource_ids:
+
+                if not resource.calendar_id:
+                    continue
+
+                if not resource.calendar_id.leave_ids:
```

Comment:
merge this if with the above if

Reply:
better, remove the above if, because it's validated when iterating on leave_ids
```python
resource.calendar_id 
False

hence: 
resource.calendar_id.leave_ids
False
```